manta-data/src/metadata.rs

```
@@ -16,47 +16,45 @@
 
 use super::*;
 use ark_std::vec::Vec;
-use manta_crypto::{HashParam, LedgerMerkleTree};
+use manta_crypto::{LeafHashParam, LedgerMerkleTree, TwoToOneHashParam};
 use manta_error::MantaError;
 
 pub trait BuildMetadata {
-	type Param;
+	type LeafParam;
+	type TwoToOneParam;
 	type Ledger;
 	type Metadata;
 
+	/// Builds `Self::Metadata` from `self` and the current state of `ledger`.
 	fn build(
 		&self,
-		param: &Self::Param,
+		leaf_param: &Self::LeafParam,
+		two_to_one_param: &Self::TwoToOneParam,
 		ledger: &Self::Ledger,
 	) -> Result<Self::Metadata, MantaError>;
 }
 
 impl BuildMetadata for MantaAsset {
-	type Param = HashParam;
+	type LeafParam = LeafHashParam;
+	type TwoToOneParam = TwoToOneHashParam;
 	type Ledger = Vec<[u8; 32]>;
 	type Metadata = SenderMetaData;
 
-	/// Build the `SenderMetaData` from sender's `MantaAsset`
-	/// and the current state of the ledger.
 	fn build(
 		&self,
-		param: &Self::Param,
+		leaf_param: &Self::LeafParam,
+		two_to_one_param: &Self::TwoToOneParam,
 		leaves: &Self::Ledger,
 	) -> Result<Self::Metadata, MantaError> {
-		let tree = LedgerMerkleTree::new(param.clone(), &leaves)?;
-		let root = tree.root();
-
-		let index = match leaves.iter().position(|x| *x == self.utxo) {
-			Some(p) => p,
-			None => {
-				return Err(MantaError::LeavesNotFound);
-			}
-		};
-		let membership = tree.generate_proof(index, &self.utxo)?;
-
+		let index = leaves
+			.iter()
+			.position(|l| *l == self.utxo)
+			.ok_or(MantaError::LeavesNotFound)?;
+		let tree = LedgerMerkleTree::new(leaf_param, two_to_one_param, leaves)?;
+		let membership = tree.generate_proof(index)?;
```

I reordered these so we can fail faster. If the leaf has no index then we have no business even constructing the tree to get a membership proof from it.